From 79355ce01a5f1373bcc501538212229a830030f6 Mon Sep 17 00:00:00 2001
From: Gerd Hoffmann <kraxel@redhat.com>
Date: Tue, 7 Feb 2017 11:18:31 +0000
Subject: [PATCH 5/8] cirrus: handle negative pitch in
 cirrus_invalidate_region()

RH-Author: Gerd Hoffmann <kraxel@redhat.com>
Message-id: <1486466314-7710-6-git-send-email-kraxel@redhat.com>
Patchwork-id: 73576
O-Subject: [RHEL-6.9 qemu-kvm PATCH 5/8] cirrus: handle negative pitch in cirrus_invalidate_region()
Bugzilla: 1418230 1418231 1419416 1419417
RH-Acked-by: Dr. David Alan Gilbert <dgilbert@redhat.com>
RH-Acked-by: Laurent Vivier <lvivier@redhat.com>
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>

From: Wolfgang Bumiller <w.bumiller@proxmox.com>

cirrus_invalidate_region() calls memory_region_set_dirty()
on a per-line basis, always ranging from off_begin to
off_begin+bytesperline. With a negative pitch off_begin
marks the top most used address and thus we need to do an
initial shift backwards by a line for negative pitches of
backward blits, otherwise the first iteration covers the
line going from the start offset forwards instead of
backwards.
Additionally since the start address is inclusive, if we
shift by a full `bytesperline` we move to the first address
*not* included in the blit, so we only shift by one less
than bytesperline.

Signed-off-by: Wolfgang Bumiller <w.bumiller@proxmox.com>
Message-id: 1485352137-29367-1-git-send-email-w.bumiller@proxmox.com

[ kraxel: codestyle fixes ]

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>
(cherry picked from commit f153b563f8cf121aebf5a2fff5f0110faf58ccb3)

[ rhel6: drop assert(), it conflicts and is pointless in the old,
         pre-memory-api code base ]

Signed-off-by: Danilo C. L. de Paula <ddepaula@redhat.com>
---
 hw/cirrus_vga.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/hw/cirrus_vga.c b/hw/cirrus_vga.c
index 68a6ef5..0ac0863 100644
--- a/hw/cirrus_vga.c
+++ b/hw/cirrus_vga.c
@@ -645,6 +645,10 @@ static void cirrus_invalidate_region(CirrusVGAState * s, int off_begin,
     int off_cur;
     int off_cur_end;
 
+    if (off_pitch < 0) {
+        off_begin -= bytesperline - 1;
+    }
+
     for (y = 0; y < lines; y++) {
 	off_cur = off_begin;
 	off_cur_end = (off_cur + bytesperline) & s->cirrus_addr_mask;
-- 
1.8.3.1

